ActivatePubKey/handler.ts

```diff
@@ -0,0 +1,216 @@
+    TE.mapLeft(domainErrorToResponseError),
+    TE.filterOrElseW(isPendingLollipopPubKey, doc => {
+      const err = `Unexpected status on pop document during activation: ${doc.status}`;
+      context.log.error(err);
+      return ResponseErrorInternal(err);
+    }),
+    TE.bindTo("popDocument"),
+    TE.bindW("assertionFileName", () =>
+      pipe(
+        `${body.fiscal_code}-${assertion_ref}`,
+        AssertionFileName.decode,
+        TE.fromEither,
+        TE.mapLeft(errors => {
+          const err = `Could not decode assertionFileName | ${readableReportSimplified(
+            errors
+          )}`;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
+        }),
+        TE.chainFirst(assertionFileName =>
+          pipe(
+            assertionWriter(assertionFileName, body.assertion),
+            // eslint-disable-next-line sonarjs/no-identical-functions
+            TE.mapLeft(error => {
+              const err = error.detail;
+              context.log.error(err);
+              return ResponseErrorInternal(err);
+            })
+          )
+        )
+      )
+    ),
+    TE.bindW("jwkPubKeyFromString", ({ popDocument }) =>
+      pipe(
+        popDocument.pubKey,
+        JwkPublicKeyFromToken.decode,
+        TE.fromEither,
+        TE.mapLeft(errors => {
+          const err = `Could not decode public key | ${readableReportSimplified(
+            errors
+          )}`;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
+        })
+      )
+    ),
+    TE.bindW("assertionRefs", ({ popDocument, jwkPubKeyFromString }) =>
+      pipe(
+        getAllAssertionsRef(
+          MASTER_HASH_ALGO,
+          getAlgoFromAssertionRef(popDocument.assertionRef),
+          jwkPubKeyFromString
+        ),
+        TE.mapLeft((error: Error) => {
+          const err = error.message;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
+        })
+      )
+    ),
+    TE.bindW(
+      "retrievedPopDocument",
+      ({ popDocument, assertionRefs, assertionFileName }) =>
+        activatePubKeyForAssertionRef(popDocumentWriter, context)(
+          assertionFileName,
+          assertionRefs.master,
+          body,
+          popDocument.pubKey
+        )
+    ),
+    TE.chain(
+      ({
+        assertionRefs,
+        assertionFileName,
+        popDocument,
+        retrievedPopDocument
+      }) =>
+        assertionRefs.used
+          ? activatePubKeyForAssertionRef(popDocumentWriter, context)(
+              assertionFileName,
+              assertionRefs.used,
+              body,
+              popDocument.pubKey
+            )
+          : TE.of(retrievedPopDocument)
+    ),
+    TE.chainW(
+      flow(
+        TE.fromPredicate(isValidLollipopPubKey, () => {
+          const err = `Unexpected retrievedPopDocument with a not VALID status`;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
```

Comment:
This block of code could be extracted into a `logAndReturnAnInternalErrorResponse(message: string, context: Context) => (err: error | undefined) => IResponseErrorInternal` function that takes a messages and return a function that takes an error to return a ResponseErrorInternal after logging that error. I think a utility function could improve readability by eliminating all these similar blocks of code for a `logAndReturnAnInternalErrorResponse("my message", context)` if there is an error then we could also stardardize the way we print our messages with the error.

Reply:
Look to https://github.com/pagopa/io-functions-lollipop/pull/13#discussion_r1119118394

Reply:
inserted a method in https://github.com/pagopa/io-functions-lollipop/pull/13/commits/b5d5d231fff8ca9e7201d2acf445b0ddaca4a880